src/wallet/wallet.cpp

```diff
@@ -922,18 +922,18 @@ CWalletTx* CWallet::AddToWallet(CTransactionRef tx, const CWalletTx::Confirmatio
         wtx.nTimeSmart = ComputeTimeSmart(wtx);
         AddToSpends(hash, &batch);
 
-        std::vector<std::pair<const CTransactionRef&, unsigned int>> outputs;
-        for(unsigned int i = 0; i < wtx.tx->vout.size(); ++i) {
+        std::set<COutPoint> candidates;
+        for (unsigned int i = 0; i < wtx.tx->vout.size(); ++i) {
             if (IsMine(wtx.tx->vout[i]) && !IsSpent(hash, i)) {
                 setWalletUTXO.insert(COutPoint(hash, i));
-                outputs.emplace_back(wtx.tx, i);
+                candidates.emplace(hash, i);
```

Comment:
std::vector usually has significantly better performance (10x), RAM usage is also over-headed (due to bunch of small allocation 32bytes per one item somewhere randomly allocated in heap, instead consequence allocation as vector) compare to std::set.

`std::set` has a great usage when you combine 2 types of actions: "add" and "request". In case if you know all elements, you should use data structure with linear memory inside for better performace.

If it is a know case here to have duplicates, consider using std::vector and remove duplicates by std::sort + std::unique + std::erase.

Reply:
Have tried to prefer `std::vector` where I can but the places that still use `std::set` do so because `setWalletUTXO` is a `std::set`.

`ListProTxCoins()` originally iterated through `setWalletUTXO` and while it was generalized to accept any input, they still are `std::set`s as it still needs to accept `setWalletUTXO`. Due to the `AddWalletUTXOs()` > `LockProTxCoins()` > `ListProTxCoins()` call chain, the arguments bubbling up and return types are also `std::set`s.

I'm not very confident about a prune/deduplication function in case we change `setWalletUTXO`'s type outright.